Zend/zend_vm_def.h

```diff
@@ -7030,6 +7030,31 @@ ZEND_VM_COLD_CONST_HANDLER(169, ZEND_COALESCE, CONST|TMP|VAR|CV, JMP_ADDR)
 	ZEND_VM_NEXT_OPCODE();
 }
 
+ZEND_VM_HOT_NOCONST_HANDLER(195, ZEND_JMP_NULL, CONST|TMPVAR|CV, JMP_ADDR)
+{
+	USE_OPLINE
+	zval *val;
+
+	val = GET_OP1_ZVAL_PTR_UNDEF(BP_VAR_R);
+
+	if (Z_TYPE_INFO_P(val) != IS_NULL) {
```

Comment:
```suggestion
	if (Z_TYPE_INFO_P(val) >= IS_NULL) {
```
So we take the other branch for IS_UNDEF. We should also be checking for IS_UNDEF there and throwing an undef var warning.

Reply:
This deliberately doesn't handle `IS_UNDEF` to let the function call or property access instructions handle the undefined variable error. Would you rather make this explicit in this handler?

Reply:
IS_UNDEF should be interpreted like IS_NULL, so the property access will not be executed in that case?

Reply:
> IS_UNDEF should be interpreted like IS_NULL, so the property access will not be executed in that case?


Not according to the RFC. `JMP_NULL` will not jump on `IS_UNDEF` values and thus cause an undefined variable warning in a later instruction.

I think this has been mentioned before. I'm open to also skipping undefined values (I've also mentioned skipping `false`) but most internals seemed to be against it. JavaScript for example skips all falsy values.

Reply:
There are three possible behaviors here:

 1. Silently ignore undefined variables and short-circuit (viable, but I would be against that).
 2. Throw an undefined variable warning and short-circuit (viable, and this is what I'm suggesting).
 3. Don't short circuit (not viable, because is violates language invariant that reading an undef variable returns null).

Unless I'm missing something, you are currently implementing option 3, while your choice is between option 1 or 2.

In any case, please add a test for undef variables.

Reply:
Yeah option 3 is definitely wrong. I haven't considered option 2.